### ISSUE TITLE
Problems with displaying chinese

### ISSUE THREAD

**Issue by chenas** [ocornut/imgui#307]
Such as the title.

**Comment by paultech** [comment]
If you read the top of https://github.com/ocornut/imgui/blob/master/imgui.cpp you will see the answer is:

 A: When loading a font, pass custom Unicode ranges to specify the glyphs to load. ImGui will support UTF-8 encoding across the board.
    Character input depends on you passing the right character code to io.AddInputCharacter(). The example applications do that.
      io.Fonts->AddFontFromFileTTF("myfontfile.ttf", size_in_pixels, NULL, io.Fonts->GetGlyphRangesJapanese());  // Load Japanese characters
      io.Fonts->GetTexDataAsRGBA32() or GetTexDataAsAlpha8()
      io.ImeWindowHandle = MY_HWND;      // To input using Microsoft IME, give ImGui the hwnd of your application

**Comment by chenas** [comment]
I have tried that, it didn`t work. I try to display chinese characters, but just english.

io.Fonts->AddFontFromFileTTF("../../extra_fonts/Cousine-Regular.ttf", 16.0f, NULL, io.Fonts->GetGlyphRangesChinese());
ImGui::Text("\u6211\u662F\u4E2D\u6587");
ImGui::Text("fsdfsdad我是中文");

The demo example that I run in widows, also can`t display non-latin characters.

**Comment by ocornut** [comment]
The Cousine font doesn't have any Asian character. Use Arial Unicode or a font that support those characters.

**Comment by chenas** [comment]
I have tried fonts below, but it doesn`t work. It shows "?????" or "口口口" instead.
I also tried to save your file as 'UTF-8 without signature' with vs2010.
io.Fonts->AddFontFromFileTTF("../../extra_fonts/arial_unicode_ms.ttf", 10.0f);
io.Fonts->AddFontFromFileTTF("../../extra_fonts/Arial.ttf", 18.0f);
io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());

**Comment by ocornut** [comment]
Only the last one here has the correct parameter io.Fonts->GetGlyphRangesChinese().
But the path is incorrect.
You need to use something like "c:\\Windows\\Fonts\\ArialUni.ttf".
Because single \ is an escape character in a string constant.

**Comment by ocornut** [comment]
Only the last one here has the correct parameter io.Fonts->GetGlyphRangesChinese(). 

Try to load only one font at a time - If you try to load more than one font you need to change font with PushFont/PopFont.

Make sure you use "c:\Windows\Fonts\ArialUni.ttf" and the file exists (you can debug inside AddFontFromFileTTF and see if the file is correctly opened for you).

`ImGui::Text("\u6211\u662F\u4E2D\u6587");` this are 16-bit unicode points, it won't work as UTF-8.
`ImGui::Text("fsdfsdad我是中文");` should work if your source file is UTF-8 and the compiler accept it.

If you are not sure about your source code encoding, first try with `ImGui::Text("Kanjis: \xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e (nihongo)");`

**Comment by chenas** [comment]
I tried what you said, that doesn`t work. 
io.Fonts->AddFontFromFileTTF("../../extra_fonts/Arial.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());
ImGui::Text("fsdfsdad我是中文dfd");
ImGui::Text("Kanjis: \xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e (nihongo)");

**Comment by ocornut** [comment]
does "../../extra_fonts/Arial.ttf" exists? It is a full unicode font?

**Comment by chenas** [comment]
I copy from C:\\Windows\\Fonts

**Comment by ocornut** [comment]
It works here. Are you copying ArialUni.ttf and NOT Arial.ttf ?
```
    ImGuiIO& io = ImGui::GetIO();
    io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());
```
[...]
```
ImGui::Text("fsdfsdad我是中文dfd");
ImGui::Text("Kanjis: \xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e (nihongo)");
```
![capture](https://cloud.githubusercontent.com/assets/8225057/9464226/1d6fb7b0-4b1a-11e5-9fff-33d889b20964.PNG)

**Comment by chenas** [comment]
What！My develope env ： win7 64bit, vs2010.
![qq 20150825181725](https://cloud.githubusercontent.com/assets/3212124/9464354/cf3c2e02-4b55-11e5-85c4-037aee5d0b44.png)
![2](https://cloud.githubusercontent.com/assets/3212124/9464394/1061b492-4b56-11e5-8495-37eebdb30c24.png)

**Comment by ocornut** [comment]
What's the size of your font file?

**Comment by chenas** [comment]
21.8M
My email is chen.guoshu@sfit.shfe.com.cn. Can you send me your ArialUni.ttf?

**Comment by ocornut** [comment]
No, we probably have the same one.
Post your program code on pastebin.com or gist.

**Comment by chenas** [comment]
I don`t think it`s the problem of the program code, I just downloaded from your git, and run in my machine. And I want to test if the Chinese characters can be displayed.

**Comment by chenas** [comment]
I downloaded ArialUni.ttf online, and added to the project.
ImGuiIO& io = ImGui::GetIO();
io.Fonts->AddFontFromFileTTF("../../extra_fonts/ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesJapanese());
It display correctly!

But below, can`t display any characters, like the first picture I uploaded before.
io.Fonts->AddFontFromFileTTF("../../extra_fonts/ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());

Anyway, thanks for helping!

**Comment by heroboy** [comment]
1. Use this font: "C:\\windows\\fonts\\SimSun.ttc" or "msyh.ttc" or "simhei.ttf".
2. You should convert string to utf8 encoding.

**Comment by chenas** [comment]
I tried, even the English can`t be displayed correctly!
io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\msyh.ttc", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());

**Comment by ocornut** [comment]
Try this:

- AddFontDefault
- then add your font

- call ImGui::ShowTestWindow() and look in "Window Options"->"Fonts"->Atlas Texture what do you see?

![atlas](https://cloud.githubusercontent.com/assets/8225057/9466614/041599a6-4b2e-11e5-89b4-c8c4992134bb.PNG)

**Comment by chenas** [comment]
io.Fonts->AddFontDefault();
io.Fonts->AddFontFromFileTTF("../../extra_fonts/ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesChinese());
![1](https://cloud.githubusercontent.com/assets/3212124/9467129/0d7bdb0c-4b6c-11e5-8c26-cb209097a582.png)

io.Fonts->AddFontDefault();
io.Fonts->AddFontFromFileTTF("../../extra_fonts/ArialUni.ttf", 18.0f, NULL, io.Fonts->GetGlyphRangesJapanese());
![2](https://cloud.githubusercontent.com/assets/3212124/9467155/32349880-4b6c-11e5-86f1-0608567fc810.png)

**Comment by ocornut** [comment]
It looks like your problem is with uploading the texture. With the default oversampling settings the texture will be very big and it may not work with your drivers. What rendering backend are you using?

Lower quality:
```
    ImFontConfig font_config; font_config.OversampleH = 1; font_config.OversampleV = 1; font_config.PixelSnapH = 1;
    io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\ArialUni.ttf", 18.0f, &font_config, io.Fonts->GetGlyphRangesChinese());
```

OversampleH is by default 3. You can try 1 or 2.

This is mainly because the atlas system is not currently dynamic..we're uploading all 22000 characters to a texture which is a big big waste here.

**Comment by chenas** [comment]
I run the opengl_exmple, suppose opengl rendering backend.
I didn`t use AddFontDefault(), some code like below：
ImGuiIO& io = ImGui::GetIO();
ImFontConfig font_config; 
font_config.OversampleH = 1; //or 2 is the same
font_config.OversampleV = 1; 
font_config.PixelSnapH = 1;
io.Fonts->AddFontFromFileTTF("../../extra_fonts/ArialUni.ttf", 18.0f, &font_config, io.Fonts->GetGlyphRangesChinese());

It seems Chinese characters can be displayed correctly, but some characters will be lost.
            ImGui::Text("hello 菜单 在哪里文件我是中文a");
            ImGui::Text("hello 菜单 在哪 里文件 我是中文a");
![1](https://cloud.githubusercontent.com/assets/3212124/9469081/adcdc254-4b77-11e5-8948-cd44926b3930.png)

**Comment by ocornut** [comment]
The second line has spaces I wonder if perhaps they have incorrect width.
What happens if you add your 2 lines twice?

            ImGui::Text("hello 菜单 在哪里文件我是中文a");
            ImGui::Text("hello 菜单 在哪 里文件 我是中文a");
            ImGui::Text("hello 菜单 在哪里文件我是中文a");
            ImGui::Text("hello 菜单 在哪 里文件 我是中文a");

**Comment by chenas** [comment]
![1](https://cloud.githubusercontent.com/assets/3212124/9469379/4218fa72-4b79-11e5-8d60-87cea12ca428.png)

By the way, the ArialUni.ttf is too big, you are right, it`s a big big waste. Do I have another way?

**Comment by ocornut** [comment]
What happen if you remove the spacing between 在哪 and 里文 ?
Maybe it isn't a regular U+0020 space and it has a wrong width (need to investigate).

You can load a limited range, by copying and modifying this function 
```
const ImWchar*  ImFontAtlas::GetGlyphRangesChinese()
{
    static const ImWchar ranges[] =
    {
        0x0020, 0x00FF, // Basic Latin + Latin Supplement
        0x3000, 0x30FF, // Punctuations, Hiragana, Katakana
        0x31F0, 0x31FF, // Katakana Phonetic Extensions
        0xFF00, 0xFFEF, // Half-width characters
        0x4e00, 0x9FAF, // CJK Ideograms
        0,
    };
    return &ranges[0];
}
```
If you know the ranges of characters you need.

Otherwise, unfortunately you can't yet. I need to rewrite ImFontAtlas to allocate characters dynamically as they are needed.

**Comment by chenas** [comment]
It displayed like the third line of the picture that I uploaded. The space is a Chinese character, too.

Unfortunately, I don't know the range of Chinese characters. I will try to find it on line later.

**Comment by ocornut** [comment]
I'd like to know which character it is (it seems like a normal space when pasted here).
Can you dump the string?
Something like

```
const char* p = "hello 菜单 在哪 里文件 我是中文a";
while (*p)
{
  printf("%02X\n", *p);
  p++;
}
```

**Comment by chenas** [comment]
I copied your code and ran.
![1](https://cloud.githubusercontent.com/assets/3212124/9481000/42a552de-4bb9-11e5-8ceb-ae07db1bbbd7.png)

Also, I find a very strange thing.
          ImGui::Text("hello 菜单 在哪 里文件 我是中文a");
If I delete an "a" at the end of the Chinese characters, the compiler will say there is something wrong with this line, seems like missing bracket ')'.
But, it`s ok as below.
	ImGui::Text("hello 菜单 在哪里文件我是中文a");
        ImGui::Text("hello 菜单 在哪 里文件 我是中文");

**Comment by ocornut** [comment]
Sorry try this one
```const char* p = "hello 菜单 在哪 里文件 我是中文a";
while (*p)
{
  printf("%02X\n", (unsigned char)*p);
  p++;
}
```
I'll look at it tomorrow :)
Thanks!

**Comment by chenas** [comment]
:)
![1](https://cloud.githubusercontent.com/assets/3212124/9482075/b17e6354-4bc2-11e5-92e8-a284aaaf928a.png)

**Comment by hypernewbie** [comment]
22K characters? so THAT'S where all the gpu vidmem went

**Comment by chenas** [comment]
@hypernewbie  Sorry, I don`t understand what you said. Could you tell me more?

**Comment by ocornut** [comment]
I think \xE4\xBB\x3F is invalid UTF-8

This page doesn't decode it: https://mothereff.in/utf-8
And the code in ImGui doesn't decode it either. Investigating.

**Comment by ocornut** [comment]
I have made a change to the UTF-8 decoder to avoid terminating a string on malformed input.

![malformed utf8 decoder](https://cloud.githubusercontent.com/assets/8225057/9520817/4ccf6772-4cc2-11e5-805f-849295f00be9.PNG)

So it is easier to understand what's going on, but it is a bug in your data/software/visual studio. Maybe if you just delete the line and input it again it will be correct?

As for VRAM correct, it is in my task list to allocate and render memory only when characters are used, but I don't know when I can do it.

**Comment by BentleyBlanks** [comment]
It seems workd after I add prefix 「u8」to the string. Looks like this

```cpp
ImGui::Text(u8"hello 菜单 在哪里文件我是中文a");  // success
ImGui::Text("hello 菜单 在哪里文件我是中文a");  // didn't workd
```

Maybe it is the problem of Visual Studio ?

**Comment by ocornut** [comment]
Using u8 if the standard c++11 way to request a string literal in utf-8. I've been starting to make that more explicit in the docs/examples.

**Comment by sgf** [comment]
@BentleyBlanks  i think thats because ur  code file is not unicode(u8 or gb2132?) format.

**Comment by fisherwise** [comment]
Facing this weird stuff today. The program just fine 2 minutes ago and I re-compile it then every character becomes white box

**Comment by wollong** [comment]
How can I support to display multi-language using format like this 
string str = u8"我是中文";
ImGui::Text("%s",str);

It's possible to make it?

**Comment by ocornut** [comment]
> It's possible to make it?

Please read FAQ and `misc/fonts/README.txt`

**Comment by wollong** [comment]
> It's possible to make it?
> 
> Please read FAQ and misc/fonts/README.txt

thx a lot! figured out codecvt works

**Comment by sweihub** [comment]
It works for me with the code for latest imgui
```
// io.Fonts->AddFontDefault();  --- DON'T CALL THIS
ImFont* font = io.Fonts->AddFontFromFileTTF("c:\\Windows\\Fonts\\msyh.ttc", 
        16.0f, NULL, io.Fonts->GetGlyphRangesChineseFull());
...

ImGui::Begin(u8"Hello, world! 这是窗口标题"); 
```

**Comment by xfdang** [comment]
It has been a while, but I slove my problem by toggle 'Use Unicode UTF-8 for worldwide language support' in region settings(win10).
![image](https://user-images.githubusercontent.com/3445440/191684349-f066f8a1-5839-4689-b64b-8d52b3629b39.png)

**Comment by PathogenDavid** [comment]
@xfdang: You really should not rely on the system codepage being UTF-8, doing so means your project won't compile correctly on other machines.

You're much better off either [setting it explicitly](https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8), using UTF-8 BOM encoding for your files, or using the `u8` prefix on strings.

**Comment by xiangfa7766** [comment]
Now everyone have a good  solution

**Comment by mlfkhf** [comment]
![image](https://github.com/user-attachments/assets/a1ee94d9-8d75-4060-bbd8-f395af2f35fa)
The picture in a tutorial on a blog my found shows this problem was out of existance clearly.
By the way, I'm sorry to say, why many library don't use std::wstring? :(

**Comment by ocornut** [comment]
FYI, work on `dynamic_fonts` branch aimed to be merged in 1.92 (#8465) simplifies this: glyphs are rasterized and packed on demand when they are used. Passing glyph ranges becomes unnecessary.
